File: scripts/save-session-for-memory.js
#!/usr/bin/env node
/**
 * Hook: SessionEnd
 * Assemble les chunks PreCompact + transcript final, extrait le contenu sémantique
 * et le sauvegarde pour Claude Mem.
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

// Add lib directory to path for imports
const libDir = path.join(__dirname, 'lib');
const debugLogger = require(path.join(libDir, 'debug-logger.js'));

const { logHookStart, logHookEnd, logDebug } = debugLogger;

const SUMMARIES_DIR = path.join(os.homedir(), '.claude', 'compacted-summaries');
const CHUNKS_DIR = path.join(os.homedir(), '.claude', 'memory-chunks');

function assembleFullTranscript(sessionId, finalTranscriptPath) {
    const allEntries = [];

    // 1. Charger les chunks PreCompact s'ils existent
    const indexPath = path.join(CHUNKS_DIR, `${sessionId}_index.json`);
    if (fs.existsSync(indexPath)) {
        try {
            const indexData = fs.readFileSync(indexPath, 'utf-8');
            const index = JSON.parse(indexData);

            // Trier par chunk_number
            const sortedIndex = index.sort((a, b) => (a.chunk_number || 0) - (b.chunk_number || 0));

            for (const chunkInfo of sortedIndex) {
                const chunkPath = path.join(CHUNKS_DIR, chunkInfo.filename);
                if (fs.existsSync(chunkPath)) {
                    try {
                        const content = fs.readFileSync(chunkPath, 'utf-8');
                        const lines = content.split('\n').filter(l => l.trim());

                        for (const line of lines) {
                            try {
                                const entry = JSON.parse(line);
                                allEntries.push(entry);
                            } catch (e) {
                                // Skip malformed lines
                            }
                        }
                    } catch (e) {
                        logDebug('SessionEnd', 'save-session-for-memory.js', `Error reading chunk ${chunkInfo.filename}: ${e.message}`, 'ERROR');
                    }
                }
            }
        } catch (e) {
            logDebug('SessionEnd', 'save-session-for-memory.js', `Error loading chunks: ${e.message}`, 'ERROR');
        }
    }

    // 2. Ajouter le transcript final
    if (fs.existsSync(finalTranscriptPath)) {
        try {
            const content = fs.readFileSync(finalTranscriptPath, 'utf-8');
            const lines = content.split('\n').filter(l => l.trim());

            for (const line of lines) {
                try {
                    const entry = JSON.parse(line);
                    allEntries.push(entry);
                } catch (e) {
                    // Skip malformed lines
                }
            }
        } catch (e) {
            logDebug('SessionEnd', 'save-session-for-memory.js', `Error loading final transcript: ${e.message}`, 'ERROR');
        }
    }

    return allEntries;
}

function extractSemanticContent(entries, cwd) {
    if (!entries || entries.length === 0) {
        return null;
    }

    const userMessages = [];
    const filesMentioned = new Set();

    for (const entry of entries) {
        // Extraire messages utilisateur
        if (entry.role === 'user') {
            const content = entry.content || '';

            if (typeof content === 'string' && content.length > 20) {
                userMessages.push(content.substring(0, 500));
            } else if (Array.isArray(content)) {
                for (const item of content) {
                    if (typeof item === 'object' && item !== null && item.type === 'text') {
                        const text = item.text || '';
                        if (text.length > 20) {
                            userMessages.push(text.substring(0, 500));
                        }
                    }
                }
            }
        }
        // Extraire fichiers mentionnés
        else if (entry.role === 'assistant') {
            const content = entry.content || '';

            if (typeof content === 'string') {
                // Chercher des patterns de fichiers (json avant js pour éviter capture partielle)
                const filePattern = /[\w/\\.-]+\.(py|json|yaml|tsx|jsx|ts|js|vue|md|css|html)/g;
                const matches = content.match(filePattern) || [];
                matches.slice(0, 20).forEach(f => filesMentioned.add(f));
            }
        }
    }

    if (userMessages.length === 0) {
        return null;
    }

    // Générer le résumé markdown
    const projectName = cwd ? path.basename(cwd) : 'Unknown';
    const now = new Date();
    const timestamp = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}-${String(now.getDate()).padStart(2, '0')} ${String(now.getHours()).padStart(2, '0')}:${String(now.getMinutes()).padStart(2, '0')}`;

    const summaryLines = [
        `# Session: ${projectName}`,
        `**Date**: ${timestamp}`,
        `**Project**: ${projectName}`,
        '',
        '## User Requests',
    ];

    for (const msg of userMessages.slice(0, 10)) {
        let cleanMsg = msg.replace(/\n/g, ' ').trim();
        if (cleanMsg.length > 200) {
            cleanMsg = cleanMsg.substring(0, 197) + '...'; // 197 + "..." = 200, then "- " + 200 = 202
        }
        summaryLines.push(`- ${cleanMsg}`);
    }

    summaryLines.push('');
    summaryLines.push('## Files Mentioned');

    if (filesMentioned.size > 0) {
        const filesArray = Array.from(filesMentioned).slice(0, 15);
        for (const f of filesArray) {
            summaryLines.push(`- ${f}`);
        }
    } else {
        summaryLines.push('- None detected');
    }

    return summaryLines.join('\n');
}

function cleanupChunks(sessionId) {
    try {
        const indexPath = path.join(CHUNKS_DIR, `${sessionId}_index.json`);
        if (fs.existsSync(indexPath)) {
            const indexData = fs.readFileSync(indexPath, 'utf-8');
            const index = JSON.parse(indexData);

            // Supprimer tous les fichiers de chunk
            for (const chunkInfo of index) {
                const chunkPath = path.join(CHUNKS_DIR, chunkInfo.filename);
                if (fs.existsSync(chunkPath)) {
                    fs.unlinkSync(chunkPath);
                }
            }

            // Supprimer l'index
            fs.unlinkSync(indexPath);
            logDebug('SessionEnd', 'save-session-for-memory.js', `Cleaned up ${index.length} chunks`, 'INFO');
        }
    } catch (e) {
        logDebug('SessionEnd', 'save-session-for-memory.js', `Error cleaning chunks: ${e.message}`, 'ERROR');
    }
}

/* istanbul ignore next */
function main() {
    logHookStart('SessionEnd', 'save-session-for-memory.js');

    let inputData = '';
    process.stdin.setEncoding('utf8');

    process.stdin.on('readable', () => {
        let chunk;
        while ((chunk = process.stdin.read()) !== null) {
            inputData += chunk;
        }
    });

    process.stdin.on('end', () => {
        try {
            const data = JSON.parse(inputData);
            const transcriptPath = data.transcript_path || '';
            const sessionId = data.session_id || 'unknown';
            const cwd = data.cwd || '';

            logDebug('SessionEnd', 'save-session-for-memory.js', `Transcript: ${transcriptPath}`, 'INFO');
            logDebug('SessionEnd', 'save-session-for-memory.js', `Session: ${sessionId}`, 'INFO');
            logDebug('SessionEnd', 'save-session-for-memory.js', `CWD: ${cwd}`, 'INFO');

            if (!transcriptPath) {
                logDebug('SessionEnd', 'save-session-for-memory.js', 'No transcript path', 'SKIP');
                logHookEnd('SessionEnd', 'save-session-for-memory.js', true);
                process.exit(0);
            }

            // Assembler tous les chunks + transcript final
            logDebug('SessionEnd', 'save-session-for-memory.js', 'Assembling full transcript...', 'INFO');
            const allEntries = assembleFullTranscript(sessionId, transcriptPath);

            if (!allEntries || allEntries.length === 0) {
                logDebug('SessionEnd', 'save-session-for-memory.js', 'No entries found', 'SKIP');
                logHookEnd('SessionEnd', 'save-session-for-memory.js', true);
                process.exit(0);
            }

            // Extraire le contenu sémantique
            logDebug('SessionEnd', 'save-session-for-memory.js', `Extracting semantic content from ${allEntries.length} entries...`, 'INFO');
            const summary = extractSemanticContent(allEntries, cwd);

            if (!summary) {
                logDebug('SessionEnd', 'save-session-for-memory.js', 'No semantic content extracted', 'SKIP');
                logHookEnd('SessionEnd', 'save-session-for-memory.js', true);
                process.exit(0);
            }

            // Sauvegarder
            if (!fs.existsSync(SUMMARIES_DIR)) {
                fs.mkdirSync(SUMMARIES_DIR, { recursive: true });
            }

            const now = new Date();
            const timestamp = `${now.getFullYear()}${String(now.getMonth() + 1).padStart(2, '0')}${String(now.getDate()).padStart(2, '0')}-${String(now.getHours()).padStart(2, '0')}${String(now.getMinutes()).padStart(2, '0')}${String(now.getSeconds()).padStart(2, '0')}`;
            const filename = `session-${timestamp}-${sessionId.substring(0, 8)}.md`;
            const filepath = path.join(SUMMARIES_DIR, filename);

            fs.writeFileSync(filepath, summary, 'utf-8');
            logDebug('SessionEnd', 'save-session-for-memory.js', `Saved: ${filename} (${allEntries.length} entries)`, 'OUTPUT');

            // Nettoyer les chunks temporaires
            cleanupChunks(sessionId);

            logHookEnd('SessionEnd', 'save-session-for-memory.js', true);
            process.exit(0);

        } catch (e) {
            logDebug('SessionEnd', 'save-session-for-memory.js', `ERROR: ${e.message}`, 'ERROR');
            logHookEnd('SessionEnd', 'save-session-for-memory.js', false);
            process.exit(0);
        }
    });
}

// Export for testing
/* istanbul ignore else */
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        assembleFullTranscript,
        extractSemanticContent,
        cleanupChunks,
        SUMMARIES_DIR,
        CHUNKS_DIR
    };
}

// Run main if executed directly
/* istanbul ignore if */
if (require.main === module) {
    main();
}

File: scripts/deprecated/save-session-for-memory.py
#!/usr/bin/env python3
"""
Hook: SessionEnd
Assemble les chunks PreCompact + transcript final, extrait le contenu sémantique
et le sauvegarde pour Claude Mem.
"""
import json
import sys
import re
from pathlib import Path
from datetime import datetime

SUMMARIES_DIR = Path.home() / ".claude" / "compacted-summaries"
CHUNKS_DIR = Path.home() / ".claude" / "memory-chunks"


def assemble_full_transcript(session_id: str, final_transcript_path: Path) -> list:
    """Assemble tous les chunks PreCompact + transcript final en une liste complète."""
    all_entries = []

    # 1. Charger les chunks PreCompact s'ils existent
    index_path = CHUNKS_DIR / f"{session_id}_index.json"
    if index_path.exists():
        try:
            with open(index_path, 'r', encoding='utf-8') as f:
                index = json.load(f)

            for chunk_info in sorted(index, key=lambda x: x['chunk_number']):
                chunk_path = CHUNKS_DIR / chunk_info['filename']
                if chunk_path.exists():
                    with open(chunk_path, 'r', encoding='utf-8') as f:
                        for line in f:
                            try:
                                entry = json.loads(line.strip())
                                all_entries.append(entry)
                            except json.JSONDecodeError:
                                pass
        except Exception as e:
            print(f"[memory] Error loading chunks: {e}", file=sys.stderr)

    # 2. Ajouter le transcript final (peut contenir des entrées après la dernière compaction)
    if final_transcript_path.exists():
        try:
            with open(final_transcript_path, 'r', encoding='utf-8') as f:
                for line in f:
                    try:
                        entry = json.loads(line.strip())
                        all_entries.append(entry)
                    except json.JSONDecodeError:
                        pass
        except Exception as e:
            print(f"[memory] Error loading final transcript: {e}", file=sys.stderr)

    return all_entries


def extract_semantic_content(entries: list, cwd: str) -> str:
    """Extrait le contenu sémantique d'une liste d'entrées de transcript."""
    if not entries:
        return None

    user_messages = []
    files_mentioned = set()

    for entry in entries:
        # Extraire messages utilisateur
        if entry.get('role') == 'user':
            content = entry.get('content', '')
            if isinstance(content, str) and len(content) > 20:
                user_messages.append(content[:500])
            elif isinstance(content, list):
                for item in content:
                    if isinstance(item, dict) and item.get('type') == 'text':
                        text = item.get('text', '')
                        if len(text) > 20:
                            user_messages.append(text[:500])
        # Extraire fichiers mentionnés
        elif entry.get('role') == 'assistant':
            content = entry.get('content', '')
            if isinstance(content, str):
                # Chercher des patterns de fichiers
                files = re.findall(r'[\w/\\.-]+\.(py|js|ts|vue|md|json|yaml|tsx|jsx|css|html)', content)
                files_mentioned.update(files[:20])

    if not user_messages:
        return None

    # Generer le resume markdown
    project_name = Path(cwd).name if cwd else "Unknown"
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M")

    summary_lines = [
        f"# Session: {project_name}",
        f"**Date**: {timestamp}",
        f"**Project**: {project_name}",
        "",
        "## User Requests",
    ]

    for msg in user_messages[:10]:
        clean_msg = msg.replace('\n', ' ').strip()
        if len(clean_msg) > 200:
            clean_msg = clean_msg[:200] + "..."
        summary_lines.append(f"- {clean_msg}")

    summary_lines.append("")
    summary_lines.append("## Files Mentioned")

    if files_mentioned:
        for f in list(files_mentioned)[:15]:
            summary_lines.append(f"- {f}")
    else:
        summary_lines.append("- None detected")

    return "\n".join(summary_lines)


def cleanup_chunks(session_id: str):
    """Supprime les chunks temporaires après assemblage."""
    try:
        index_path = CHUNKS_DIR / f"{session_id}_index.json"
        if index_path.exists():
            with open(index_path, 'r', encoding='utf-8') as f:
                index = json.load(f)

            # Supprimer tous les fichiers de chunk
            for chunk_info in index:
                chunk_path = CHUNKS_DIR / chunk_info['filename']
                if chunk_path.exists():
                    chunk_path.unlink()

            # Supprimer l'index
            index_path.unlink()
            print(f"[memory] Cleaned up {len(index)} chunks", file=sys.stderr)

    except Exception as e:
        print(f"[memory] Error cleaning chunks: {e}", file=sys.stderr)


def main():
    try:
        input_data = json.load(sys.stdin)
        transcript_path_str = input_data.get('transcript_path', '')
        session_id = input_data.get('session_id', 'unknown')
        cwd = input_data.get('cwd', '')

        if not transcript_path_str:
            sys.exit(0)

        transcript_path = Path(transcript_path_str)

        # Assembler tous les chunks + transcript final
        all_entries = assemble_full_transcript(session_id, transcript_path)

        if not all_entries:
            sys.exit(0)

        # Extraire le contenu sémantique
        summary = extract_semantic_content(all_entries, cwd)
        if not summary:
            sys.exit(0)

        # Sauvegarder
        SUMMARIES_DIR.mkdir(parents=True, exist_ok=True)
        timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
        filename = f"session-{timestamp}-{session_id[:8]}.md"
        filepath = SUMMARIES_DIR / filename

        filepath.write_text(summary, encoding='utf-8')
        print(f"[memory] Saved: {filename} ({len(all_entries)} entries)", file=sys.stderr)

        # Nettoyer les chunks temporaires
        cleanup_chunks(session_id)

        sys.exit(0)
    except Exception as e:
        print(f"[memory] Error: {e}", file=sys.stderr)
        sys.exit(0)


if __name__ == '__main__':
    main()
